specification/dartLangSpec.tex

```diff
@@ -17698,7 +18090,7 @@ \subsubsection{Asynchronous For-in}
 % This error can occur due to implicit casts and null.
 It is a dynamic type error if $o$ is not an instance of
 a class that implements \code{Stream}.
-It is a compile-time error if $D$ is empty and \id{} is a final variable.
+It is a compile-time error if $D$ is empty and \id{} is an immutable variable.
```

Comment:
... is an immutable and non-`late` variable that has potentially been assigned, or if it is immutable and `late` and definitely assigned.

Should we have an abstraction?
> A local variable *may be assigned* if it is not immutable,
> or if it's immutable and definitely unassigned,
> or it's immutable and `late` and not definitely assigned.

(Variables with initializer expressions are always definitely assigned, even if they are late and not initialized yet.)

Just so we don't have to remember every case every time.

The following are both valid and run today:
```dart
  final int x;  
  for (x in [1]) {print(x);}
```
```dart
  late final int x;  
  for (x in [1]) {}
  print(x);
```

Reply:
I get an error in DartPad for this:
```dart
void main() {
  final int x;
  for (x in [1]) {
    print(x);
  }
}
```
That would make sense, too, because there is no analysis which will establish that the for-statement iterates exactly one time. Also, it matches the treatment of the desugaring as a while-statement, where we are also not attempting to prove that it will execute the body at most once. I get the compile-time error for this one as well:

```dart
void main() {
  late final int x;
  for (x in [1]) {}
  print(x);
}
```

It looks like it is consistent with the implemented static analysis to say that it is an error when there is no declaration (so we're running a for-statement with an existing variable as the iteration variable), and that variable is immutable, late or non-late. The error messages could be improved, but we do flag these situations as errors.

So how did you run those two examples without encountering any errors?

Reply:
The
```
line 8 * The late final local variable is already assigned.
```
looks spurious. At the assignment to `x2` in the loop, it's at most *possibly assigned*, and since it's `late`, that's OK.

That is, there is a possible execution sequence where the `late` variable doesn't throw when accessed, so it would be wrong to consider the program incorrect. (Non-late variable uses must be valid on *all* execution paths, late variable uses must be possibly valid on at least *one*.)

Reply:
The error message for line 8 is of course less than ideal: It does not include the name of the variable.

I don't think we can use ad-hoc reasoning and conclude that a given late variable is possibly assigned, etc, because the actual analysis will do something specific (for loops, and for any other construct), and the resulting data is used to determine where/why to emit compile-time errors. We just have to live with the outcome, even when it is "easy to see" that some other result would have been more precise and still sound.

For example, the assignment to `x` in `for (x in ...)` is presumably approximated to run at least twice (even though it might just run once or not at all). This implies that we have approximated and then concluded that there is an assignment to `x` at a time where it is definitely assigned. That is indeed true for any loop execution that iterates more than one time.

However, I do think we should take this opportunity to double check that the implementations are consistent with the specification.

So we _should_ get an error for an `x` which is immutable, both when it is non-late and when it is late, and then both when it is definitely unassigned and when it is possibly assigned. We do get that:

```dart
void main() async {
  {
    // Immutable, non-late.
    final int x;
    await for (x in Stream.fromIterable([1])) {} // Error: can only set once.
    print(x); // Error: potentially unassigned: OK.
  }
  {
    // Immutable, late, definitely unassigned at `for`.
    late final int x;
    await for (x in Stream.fromIterable([1])) {} // Error: is already assigned.
    print(x);
  }
  {
    // Immutable, late, possibly assigned at `for`.
    late final int x;
    if (2 > 1) x = 1;
    await for (x in Stream.fromIterable([1])) {} // Error: is already assigned.
    print(x);
  }
}
```

I think we should just keep the spec as it is, given that this is also the implemented behavior.